woocommerce/woocommerce-ios
# Dynamic Dashboard: Coupons card

Closes: #12716 
Closes: #12728
Part of: #12698



## Description


Completes the UI and logic for the coupons card in dashboard.

Changes
- Action and remote method for loading coupons using coupon IDs.
- Fetch coupon report and coupons from remote and display those.
- Empty view and content unavailable view.
- Unit tests.

## Testing instructions


- Enable the feature flag `dynamicDashboardM2` and build the app.
- Enable the Coupon card on the dashboard.
- Confirm that the card now displays the 3 most active coupons for the selected time range. 
- Validate that the UI matches the design.
- Change the time range and validate the coupons are listed based on the selected time range.
- If there are no coupons available for a selected time range an empty view with the message "No coupon usage during this period" should be displayed.
- Tap on any coupon and validate that the coupon details view is displayed.
- Try editing or deleting the coupon and ensure that it works as expected. 
- Navigate back to the dashboard. Tap View all coupons and confirm that the coupons list view is displayed.
- Disconnect the Internet connect and reload the card. Confirm that the error state is displayed.
- Disable "WooCommerce Analytics" toggle from your wp-admin page. Reload the card and ensure that error view is displayed asking you to check WooCommerce version/settings or contact support.
- Tap the ellipsis button and dismiss the card. Confirm that event `dynamic_dashboard_hide_card_tapped` is tracked with type `coupons`.

## Screenshots


<img src="https://github.com/woocommerce/woocommerce-ios/assets/524475/e6f4ff52-a3b2-4bcd-a7a4-d2b9d9d680b6" width="320"/>


Error cases
| WooCommerce Analytics disabled | No active coupons available for the selected time range | No internet  |
|--------|--------|--------|
| ![image](https://github.com/woocommerce/woocommerce-ios/assets/524475/c3fa650c-a076-447d-8808-0d600800e9f7) | ![image](https://github.com/woocommerce/woocommerce-ios/assets/524475/57aca4dd-7788-4f65-86de-3c78690d46ba) | ![image](https://github.com/woocommerce/woocommerce-ios/assets/524475/fcdcd1ec-e63b-412d-88a6-c5788d527169) | 

wp-admin page "Enable WooCommerce Analytics" toggle

<img width="1230" alt="Screenshot 2024-05-16 at 3 16 36 PM" src="https://github.com/woocommerce/woocommerce-ios/assets/524475/4f47b693-4754-4da3-87c9-ad3aca63625a">


---
